pkg/webhook/resourcedeletionprotection/validating.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2023 The Karmada Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcedeletionprotection
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// ValidatingAdmission validates CronFederatedHPA object when creating/updating.
+type ValidatingAdmission struct {
+	Decoder *admission.Decoder
+}
+
+// Check if our ValidatingAdmission implements necessary interface
+var _ admission.Handler = &ValidatingAdmission{}
+
+// If a user assigns the ResourceDeletionProtectionLabelKey label to a specific resource,
+// and the value of this label is ResourceDeletionProtectionAlways, then deletion requests
+// for this resource will be denied.
+const (
+	ResourceDeletionProtectionLabelKey = "karmada.io/deletion-protected"
+	ResourceDeletionProtectionAlways   = "Always"
+	ResourceDeletionProtectionNever    = "Never"
+)
+
+// Handle implements admission.Handler interface.
+// It yields a response to an AdmissionRequest.
+func (v *ValidatingAdmission) Handle(_ context.Context, req admission.Request) admission.Response {
+	if req.Operation == admissionv1.Delete {
+		// Parse the uncertain type resource object
+		obj := &unstructured.Unstructured{}
+		if err := v.Decoder.DecodeRaw(req.OldObject, obj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		klog.V(2).Infof("Validating ResourceDeletionProtection for resource: Kind:%s Name:%s", req.Kind.Kind, obj.GetName())
+
+		if value, ok := obj.GetLabels()[ResourceDeletionProtectionLabelKey]; ok {
+			// In normal, requests will be processed here.
+			// Only ResourceDeletionProtectionAlways value will be denied
+			if value == ResourceDeletionProtectionAlways {
+				return admission.Denied(fmt.Sprintf("Resource are protected, remove the label '%s' for delete the resource", ResourceDeletionProtectionLabelKey))
+			}
+			return admission.Allowed("")
+		} else {
+			// The resource must have the label, the webhook configuration will filter resources based on this label
+			return admission.Errored(http.StatusBadRequest, fmt.Errorf("the resource should have label '%s'", ResourceDeletionProtectionLabelKey))
```

Comment:
If the target label is not included, can we skip it directly?

Reply:
sure!